components/eam/src/physics/cam/shoc.F90

```diff
@@ -3482,49 +3458,46 @@ subroutine vd_shoc_solve(&
   integer, intent(in) :: shcol
   ! number of mid-point levels
   integer, intent(in) :: nlev
-  ! number of levels on the interface
-  integer, intent(in) :: nlevi
   ! superdiagonal
-  real(rtype), intent(in) :: ca(shcol,nlev)
+  real(rtype), intent(in) :: du(shcol,nlev)
   ! subdiagonal
-  real(rtype), intent(in) :: cc(shcol,nlev)
-  ! 1./(1.+ca(k)+cc(k)-cc(k)*ze(k-1))
-  real(rtype), intent(in) :: denom(shcol,nlev)
-  ! Term in tri-diag. matrix system
-  real(rtype), intent(in) :: ze(shcol,nlev)
+  real(rtype), intent(in) :: dl(shcol,nlev)
+  ! diagonal
+  real(rtype), intent(in) :: d(shcol,nlev)
 
 ! IN/OUT VARIABLES
   real(rtype), intent(inout) :: var(shcol,nlev)
 
 ! LOCAL VARIABLES
   integer :: i, k
-  ! Term in tri-diag solution
-  real(rtype) :: zf(shcol,nlev)
+  ! Temporary variables for tridiag solve
+  real(rtype) :: temp_dl(shcol,nlev)
+  real(rtype) :: temp_d(shcol,nlev)
 
+  ! Copy over data to temporary variables
+  temp_dl(:,:) = dl(:,:)
+  temp_d (:,:) = d (:,:)
 
-  ! Calculate zf(k). Terms zf(k) and ze(k) are required in solution of
-  ! tridiagonal matrix defined by implicit diffusion equation.
-  ! Note that only levels ntop through nbot need be solved for.
-
-  do i=1,shcol
-    zf(i,nlev) = var(i,nlev) * denom(i,nlev)
+  ! Compute Thomas factorization
```

Comment:
I think this factorization block can be moved to the vd_shoc_decomp routine. Overwrite dl and d there. Then these can be used in vd_shoc_solve repeatedly without temp_dl,d. That should be more efficient (save redundant work) and mimic what will happen in the C++ code.

Reply:
Good point! I moved it.